ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,899 @@
+        }
+        *ptrs++ = NULL;
+
+        // Copy all mappings, we need to change them.
+        info->mappings = mappings;
+        for (int i = 0; i < 3; ++i) {
+            mapping *m = mappings++;
+            m->uuid = NULL;
+            m->current_fd = standard_fds[i];
+            m->wanted_fd = i;
+        }
+        FOREACH_LIST(map, id_mapping) {
+            mapping *m = mappings++;
+            value item = Field(map, 0);
+            m->uuid = append_string(&strings, String_val(Field(item, 0)));
+            m->current_fd = Int_val(Field(item, 1));
+            m->wanted_fd = -1;
+        }
+
+#ifdef FORKEXECD_DEBUG_LOGS
+        if (strings - (char*)info != total_size
+            || initial_mappings != (mapping*) ptrs
+            || initial_strings != (char*) mappings) {
+            log_fail("internal error copying");
+            caml_stat_free(info);
+            mapped_logs_close(logs);
+            unix_error(EACCES, "safe_exec", Nothing);
+        }
+#endif
+    }
+
+    // Release Ocaml engine, it could feel all operations are pretty quick
+    // so there's no need but potentially launching a process can take
+    // some time (like files on NFS), so better safe than sorry.
+    caml_enter_blocking_section();
+    {
+    // make sure we don't use Ocaml variables anymore in the section
+    enum {
+        environment, stdin_fd, stdout_fd, stderr_fd, waiter,
+        id_mapping, syslog_stdout, redirect, args,
+        key, strings_size, res
+    };
+
+    sigset_t sigset, old_sigset;
+    int cancellation_state;
+
+    // Compute the file operations we need to do for the file mappings,
+    // this is done before vfork() call to reduce operations done between
+    // vfork() and execve().
+    int num_operations =
+        redirect_mappings(info->mappings, num_mappings, info->operations);
+
+    if (FORKEXECD_DEBUG_LOGS) {
+        for (size_t n = 0; info->args[n]; ++n)
+            log("arg %zd %s", n, info->args[n]);
+    }
+
+    // Rename all command line.
+    adjust_args(info->args, info->mappings, num_mappings);
+
+    if (FORKEXECD_DEBUG_LOGS) {
+        for (size_t n = 0; info->args[n]; ++n)
+            log("arg %zd %s", n, info->args[n]);
+    }
+
+    // Disable cancellation to avoid some signals.
+    // Glibc use some signals to handle thread cancellation.
+    pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancellation_state);
+
+    // Block all possible signals to avoid receiving some in the child.
+    // Signal mask is inherited to new process/thread will start with
+    // all signals disabled and we can safely change them.
+    sigfillset(&sigset);
+    pthread_sigmask(SIG_BLOCK, &sigset, &old_sigset);
+
+    log("vfork");
+
+    // fork
+    pid = vfork();
+    if (pid < 0) {
+        int err = errno;
+        // error, notify with an exception
+        log_fail("vfork %d", err);
+        caml_stat_free(info);
+        mapped_logs_close(logs);
+        unix_error(err, "safe_exec", Nothing);
```

Comment:
This won't work, the OCaml runtime lock was already released, calling this function will raise an exception and execute OCaml code again without taking the runtime lock first, which will corrupt the OCaml heap when run in a multithreaded program.

Reply:
Good catch, adding missing unlock.